js/src/components/tree-select-control/options.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * External dependencies
+ */
+import { CheckboxControl, Flex } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { Icon, chevronUp, chevronDown } from '@wordpress/icons';
+import classnames from 'classnames';
+
+/**
+ * This component renders a list of options and its children recursively
+ *
+ * @param {Object} params Component parameters
+ * @param {Option[]} params.options List of options to be rendered
+ * @param {string[]} params.value List of selected values
+ * @param {string} params.parent The parent of these set of options
+ * @param {Function} params.onChange Callback when an option changes
+ */
+const Options = ( {
+	options = [],
+	value = [],
+	parent = '',
+	onChange = () => {},
+} ) => {
+	const [ expanded, setExpanded ] = useState( [] );
+	/**
+	 * Returns true if all the children for the parent are selected
+	 *
+	 * @param {Option} option The parent option to check
+	 */
+	const isParentSelected = ( option ) => {
+		if ( ! option.children ) {
+			return false;
+		}
+
+		return option.children.every(
+			( child ) =>
+				value.includes( child.value ) || isParentSelected( child )
+		);
+	};
+
+	const toggleExpanded = ( option ) => {
+		setExpanded(
+			expanded.includes( option )
+				? expanded.filter( ( el ) => option !== el )
+				: [ ...expanded, option ]
+		);
+	};
+
+	return options.map( ( option ) => {
+		const isExpanded = expanded.includes( option.value );
+		const hasChildren = !! option.children?.length;
+
+		return (
+			<div
+				key={ `${ parent }-${ option.value }` }
+				role={ hasChildren ? 'treegroup' : 'treeitem' }
+				aria-expanded={ hasChildren ? isExpanded : undefined }
+			>
+				<Flex
+					className="woocommerce-tree-select-control__node"
+					justify="flex-start"
+				>
+					<Icon
+						className={ classnames(
+							'woocommerce-tree-select-control__expander-icon',
+							! hasChildren ? 'is-hidden' : false
+						) }
+						icon={ isExpanded ? chevronUp : chevronDown }
+						onClick={ () => {
+							toggleExpanded( option.value );
+						} }
+					/>
+
+					<CheckboxControl
+						className={ 'woocommerce-tree-select-control__option' }
+						value={ option.value }
+						label={ option.label }
+						checked={
+							value.includes( option.value ) ||
+							isParentSelected( option )
+						}
+						onChange={ ( checked ) => {
+							onChange( checked, option );
+						} }
+					/>
+				</Flex>
+
+				{ hasChildren && isExpanded && (
+					<div className="woocommerce-tree-select-control__children">
+						<Options
+							parent={ option.value }
+							options={ option.children }
+							onChange={ onChange }
+							value={ value }
+						/>
+					</div>
+				) }
```

Comment:
❓ I have doubts if we should render all the nodes. And hide them with CSS instead of conditionally rendering them.

I think we should render them and hide them by CSS because of the search functionality.

Reply:
💡  I used 2000 children options in one parent option to do a basic performance measure for four actions:
- 📌  Open dropdown
- 📌  Expand a long group
- 📌  Eollapse a long group
- 📌  Close dropdown 

## Conditionally control the rendering (the current implementation)

https://user-images.githubusercontent.com/17420811/157640079-e2944b2f-5a7c-42cb-b77e-7eac03d6abff.mp4

### With the current implementation, we can observe the following things from the basic metrics:

1. Actions of opening dropdown, closing dropdown, and collapsing a long group are fast enough. Even though there are some dropped frames in 60 FPS, it should be ok.
2. Expanding a long group took 1935 ms (Scripting + Rendering + Painting), and there are noticeable continuous dropped frames.
   ![2022-03-10 18 35 33](https://user-images.githubusercontent.com/17420811/157644282-ff030625-186e-454a-b483-4594bc41f46e.png)
   ![2022-03-10 18 07 44](https://user-images.githubusercontent.com/17420811/157639955-86bfd95e-4246-4ced-84de-199d5568739f.png)

## Conditionally control the `display` style by toggling a class

https://user-images.githubusercontent.com/17420811/157640712-88eb6d72-4bdb-4964-93bb-1a240c036cf7.mp4

### The observations:

1. Except for closing dropdown, the rest are significantly slower than the above approach.
2. Opening dropdown took 958 ms and a lot of dropped frames. Most time blocked by scripting.
   ![2022-03-10 18 38 22](https://user-images.githubusercontent.com/17420811/157645400-a7465b3b-a59b-4fd7-b56e-8ac6390f5786.png)
3. Expanding a long group took 1561 ms. It looks a little faster, but the total blocking time of the screen is about 3 seconds.
   ![2022-03-10 18 39 54](https://user-images.githubusercontent.com/17420811/157645453-7b631ab7-e9cb-4445-86cf-7edaeb387327.png)
4. Eollapsing a long group took 612 ms. Similar to the case of opening dropdown. Most time blocked by scripting.
   ![2022-03-10 18 42 03](https://user-images.githubusercontent.com/17420811/157645475-9dbda940-a83c-454e-b261-ba1fd07a9774.png)

## My current conclusion

These are rough observations, but it should be able to find out that in general the current practice is relatively good. Whether scripting blocking could be improved, I think we could consider it when we want to move this component forward with upstream packages.

Reply:
That helped Huge thanks